src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java

```diff
@@ -731,29 +731,26 @@ private boolean includeSelfAsArg(
   /**
    * Call a method depending on the type of an object it is called on.
    *
-   * @param positionals The first object is expected to be the object the method is called on.
+   * @param positionalArgs positional arguments to pass to the method
    * @param call the original expression that caused this call, needed for rules especially
    */
   private Object invokeObjectMethod(
+      Object value,
       String method,
-      ImmutableList<Object> positionals,
+      @Nullable BaseFunction function,
+      ImmutableList<Object> positionalArgs,
       ImmutableMap<String, Object> keyWordArgs,
       FuncallExpression call,
       Environment env)
       throws EvalException, InterruptedException {
     Location location = call.getLocation();
-    Object value = positionals.get(0);
-    BaseFunction function = Runtime.getBuiltinRegistry().getFunction(value.getClass(), method);
-    Object fieldValue =
-        (value instanceof ClassObject) ? ((ClassObject) value).getValue(method) : null;
-    ImmutableList<Object> positionalArgs =
-        includeSelfAsArg(value, function)
-            ? positionals
-            : positionals.subList(1, positionals.size());
-
     if (function != null) {
       return function.call(positionalArgs, keyWordArgs, call, env);
-    } else if (fieldValue != null) {
+    }
+    @Nullable
+    Object fieldValue =
+        (value instanceof ClassObject) ? ((ClassObject) value).getValue(method) : null;
+    if (fieldValue != null) {
```

Comment:
I think it can be simplified further:
```java
if (value instanceof ClassObject) {
  Object fieldValue = ...
```

Reply:
Oops, `.getValue()` may return null. Forget it.

Reply:
yep, I also thought about it, but I'd have to duplicate a null check anyways below because `getValue` can return `null` :(